ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -23,123 +29,204 @@ import           Network.Socket hiding (recv, recvFrom, send, sendTo)
+--
+socketAsMuxBearer
+  :: forall ptcl.
+     Mx.ProtocolEnum ptcl
+  => Socket
+  -> IO (MuxBearer ptcl IO)
+socketAsMuxBearer sd = do
+      mxState <- atomically $ newTVar Mx.Larval
+      return $ Mx.MuxBearer {
+          Mx.read    = readSocket,
+          Mx.write   = writeSocket,
+          Mx.close   = closeSocket,
+          Mx.sduSize = sduSize,
+          Mx.state   = mxState
+        }
+    where
+      readSocket :: (HasCallStack) => IO (Mx.MuxSDU ptcl, Time IO)
+      readSocket = do
+          hbuf <- recvLen' 8 []
+          --say "read"
+          --hexDump hbuf ""
+          case Mx.decodeMuxSDUHeader hbuf of
+              Left  e      -> throwM e
+              Right header -> do
+                  --say $ printf "decoded mux header, goint to read %d bytes" (Mx.msLength header)
+                  blob <- recvLen' (fromIntegral $ Mx.msLength header) []
+                  ts <- getMonotonicTime
+                  --hexDump blob ""
+                  return (header {Mx.msBlob = blob}, ts)
+
+      recvLen' :: Int64 -> [BL.ByteString] -> IO BL.ByteString
+      recvLen' 0 bufs = return $ BL.concat $ reverse bufs
+      recvLen' l bufs = do
+          buf <- Socket.recv sd l
+          if BL.null buf
+              then throwM $ Mx.MuxError Mx.MuxBearerClosed "Socket closed when reading data" callStack
+              else recvLen' (l - fromIntegral (BL.length buf)) (buf : bufs)
+
+      writeSocket :: Mx.MuxSDU ptcl -> IO (Time IO)
+      writeSocket sdu = do
+          --say "write"
+          ts <- getMonotonicTime
+          let sdu' = sdu { Mx.msTimestamp = Mx.RemoteClockModel $ fromIntegral $ ts .&. 0xffffffff }
+              buf = Mx.encodeMuxSDU sdu'
+          --hexDump buf ""
+          Socket.sendAll sd buf
+          return ts
+
+      closeSocket :: IO ()
+      closeSocket = close sd
+
+      sduSize :: IO Word16
+      sduSize = do
+          -- XXX it is really not acceptable to call getSocketOption for every SDU we want to send
+          mss <- getSocketOption sd MaxSegment
+          -- 1260 = IPv6 min MTU minus TCP header, 8 = mux header size
+          return $ fromIntegral $ max (1260 - 8) (min 0xffff (15 * mss - 8))
+
 
 hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+-- |
+-- Like @'runNetworkNodeWithSocket'@ but it allows to run an action when an
+-- exception is raised by the mux layer.  This is useful in tests.
+--
+runNetworkNodeWithSocket'
+  :: forall ptcl.
+     ( Mx.ProtocolEnum ptcl
+     , Ord ptcl
+     , Enum ptcl
+     , Bounded ptcl
+     )
+  => NetworkInterface ptcl AddrInfo IO
+  -> Maybe (Maybe SomeException -> IO ())
+  -> IO (NetworkNode AddrInfo IO)
+runNetworkNodeWithSocket' NetworkInterface {nodeAddress, knownVersions, protocols} k = do
+      (sd, hdl) <- startNode
+      return $ NetworkNode {
+          connectTo,
+          killNode  = closeConn sd hdl
```

Comment:
What is the expected meaning of `killNode`? Should it cancel all mux connection threads and wait for them to finish? I'd think so. Should also guarantee that all of those sockets are closed up properly. AFAICT this definition doesn't do that. It'll just kill the `accept` loop and leave a bunch of dangling connection threads.

Reply:
if you use both `withNetworkNode` and inside of it `withConnection` you will not be left with any dangling connection (socket and thread).  This way we don't need to track all the connections that were accepted and are running inside of `mux`.

Reply:
What about connections that were remotely-initiated?

Reply:
There is a `watcher` which will handle [this]](https://github.com/input-output-hk/ouroboros-network/pull/359/files#diff-37c626674753d27fd06d21f7c2439ec6R157). It awaits until the thread that runs the application will finish or throw, and close the socket.